src/browser/telemetry.js

```diff
@@ -640,111 +640,148 @@ class Instrumenter {
   };
 
   instrumentDom() {
-    if (!('addEventListener' in this._window || 'attachEvent' in this._window)) {
-      return;
-    }
-    const clickHandler = this.handleClick.bind(this);
-    const blurHandler = this.handleBlur.bind(this);
-    this.addListener('dom', this._window, 'click', clickHandler, true);
+    const self = this;
+    this.addListener('dom', this._window, ['click', 'dblclick', 'contextmenu'], (e) => this.handleEvent('click', e));
     this.addListener(
       'dom',
       this._window,
-      'blur',
-      blurHandler,
-      true,
+      ['dragstart', 'dragend', 'dragenter', 'dragleave', 'drop'],
+      (e) => this.handleEvent('dragdrop', e)
     );
+    this.addListener('dom', this._window, ['blur', 'focus'], (e) => this.handleEvent('focus', e));
+    this.addListener('dom', this._window, ['submit', 'invalid'], (e) => this.handleEvent('form', e));
+    this.addListener('dom', this._window, ['input', 'change'], (e) => this.handleEvent('input', e));
+    this.addListener('dom', this._window, ['resize'], (e) => this.handleEvent('resize', e));
   }
 
-  handleClick(evt) {
+  handleEvent(name, evt) {
     try {
-      const e = domUtil.getElementFromEvent(evt, this._document);
-      const hasTag = e && e.tagName;
-      const anchorOrButton =
-        domUtil.isDescribedElement(e, 'a') ||
-        domUtil.isDescribedElement(e, 'button');
-      if (
-        hasTag &&
-        (anchorOrButton ||
-          domUtil.isDescribedElement(e, 'input', ['button', 'submit']))
-      ) {
-        this.captureDomEvent('click', e);
-      } else if (domUtil.isDescribedElement(e, 'input', ['checkbox', 'radio'])) {
-        this.captureDomEvent('input', e, e.value, e.checked);
-      }
+      return {
+        click: this.handleClick,
+        dragdrop: this.handleDrag,
+        focus: this.handleFocus,
+        form: this.handleForm,
+        input: this.handleInput,
+        resize: this.handleResize,
+      }[name].call(this, evt);
     } catch (exc) {
-      // TODO: Not sure what to do here
+      console.log(`${name} handler error`, evt, exc, exc.stack);
```

Comment:
The cool thing about having a type-theoretic system, this state would become flat-out unrepresentable.